### PR TITLE
Assignment 1

### PR DESCRIPTION
**Pull request by shaurya-src**

### CODE REVIEW

File: Assignment-1/bin/brute2.dart [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    try {
+      if (!possibleTypes.contains(itemType)) {
+        throw Exception("Invalid Item Type.");
+      }
+    } catch (error) {
+      print("Invalid Item Type.");
+      errorFlag = true;
+    }
+
+    // ############################################
```
Comment (shaurya-src) [inline Assignment-1/bin/brute2.dart:166]: f

**Review by Akanksha-nuclei — CHANGES_REQUESTED** [review]

File: Assignment-1/bin/brute2.dart [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+// helper function to throw exception when value is less than or equal to zero
+void isGreaterThanZero(num val) {
+  if (val <= 0) {
+    throw Exception("Value must be greater than or equal to zero.");
+  }
+}
+
+void main() {
```
[the remark this answers is not in the document]
Comment (shaurya-src) [reply Assignment-1/bin/brute2.dart]: Moved main function to main.dart (and kept the Item class in item.dart file).

File: Assignment-1/bin/brute2.dart [hunk 3]
```diff
@@ -0,0 +1,186 @@
+import 'dart:io';
+
+class Item {
+  // Define attributes of the Item Class
+  String name;
+  double price;
+  int quantity;
+  String type;
+
```
[the remark this answers is not in the document]
Comment (shaurya-src) [reply Assignment-1/bin/brute2.dart]: Implemented enum to validate item type.

File: Assignment-1/bin/brute2.dart [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+// helper function to throw exception when value is less than or equal to zero
+void isGreaterThanZero(num val) {
+  if (val <= 0) {
+    throw Exception("Value must be greater than or equal to zero.");
+  }
+}
+
+void main() {
+  List<String> possibleTypes = ["raw", "manufactured", "imported"];
+
```
[the remark this answers is not in the document]
Comment (shaurya-src) [reply Assignment-1/bin/brute2.dart]: Implemented enum to validate item type.

File: Assignment-1/bin/brute2.dart [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    stdout.write("Enter the item name:\t");
+    String itemName = stdin.readLineSync().toString();
+
+    double itemPrice = -1;
+    int itemQuantity = -1;
+    bool errorFlag = false;
+
+    stdout.write("Enter the item price:\t");
+    try {
```
[the remark this answers is not in the document]
Comment (shaurya-src) [reply Assignment-1/bin/brute2.dart]: Done.

File: Assignment-1/bin/brute2.dart [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+  do {
+    // ############################################
+
+    stdout.write("Enter the item name:\t");
+    String itemName = stdin.readLineSync().toString();
+
+    double itemPrice = -1;
+    int itemQuantity = -1;
+    bool errorFlag = false;
```
[the remark this answers is not in the document]
Comment (shaurya-src) [reply Assignment-1/bin/brute2.dart]: Error

If I don't initialize this, I'm getting this error

File: Assignment-1/bin/brute2.dart [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+void main() {
+  List<String> possibleTypes = ["raw", "manufactured", "imported"];
+
+  // Continue taking inputs as long as user enters y/Y when asked.
+
+  do {
+    // ############################################
+
+    stdout.write("Enter the item name:\t");
+    String itemName = stdin.readLineSync().toString();
```
[the remark this answers is not in the document]
Comment (shaurya-src) [reply Assignment-1/bin/brute2.dart]: Added checks for numeric and null values.

File: Assignment-1/bin/brute2.dart [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+      // Tax calculation for IMPORTED items
+      case "imported":
+        double importDuty = 0.1 * price;
+        //
+        // not clear in the question
+        //
+        double baseTax = 0.125 * price;
+        double lumpSumAmount = price + importDuty + baseTax;
+
+        var surcharge;
```
[the remark this answers is not in the document]
Comment (shaurya-src) [reply Assignment-1/bin/brute2.dart]: Done

File: Assignment-1/lib/main.dart [hunk 9]
```diff
@@ -0,0 +1,91 @@
+import 'dart:io';
+import 'item.dart';
+
+// helper function to throw exception when value is less than or equal to zero
+void isGreaterThanZero(num val) {
+  if (val <= 0) {
+    throw Exception("Value must be greater than zero.");
+  }
+}
+
```
Comment (Akanksha-nuclei) [inline Assignment-1/lib/main.dart:10]: create a separate util class for this

File: Assignment-1/bin/brute2.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (shaurya-src) [reply Assignment-1/bin/brute2.dart]: Made the changes!

File: Assignment-1/lib/main.dart (hunk 9 above)
Comment (shaurya-src) [reply Assignment-1/lib/main.dart]: Done. Made Util class in util.dart.

File: Assignment-1/lib/item.dart [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+  Item(this.name, {this.price = 0, this.quantity = 1, required this.type});
+
+  // define late variables to store tax and final price
+  late double _salesTax = calculateSalesTax();
+  late double _finalPrice = getFinalPrice();
+
+  // Function to calculate sales tax according to item type
+  double calculateSalesTax() {
+    double _totalTax = -1;
+
```
Comment (Akanksha-nuclei) [inline Assignment-1/lib/item.dart:25]: try to remove this variable and directly return from switch statement itself

File: Assignment-1/lib/item.dart [hunk 11]
```diff
@@ -0,0 +1,112 @@
+enum ItemType { raw, manufactured, imported }
+
+class Item {
+  // Define attributes of the Item Class
+  String name;
+  double price;
+  int quantity;
+  ItemType type;
+
```
Comment (Akanksha-nuclei) [inline Assignment-1/lib/item.dart:9]: these fields can be made final

File: Assignment-1/lib/item.dart [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+  /* 
+  Constructor to initialize the object
+  name is a mandatory positional argument
+  price, quantity, type are named args
+  out of these three, only type is mandatory
+  */
+  Item(this.name, {this.price = 0, this.quantity = 1, required this.type});
+
```
Comment (Akanksha-nuclei) [inline Assignment-1/lib/item.dart:17]: check if you can make it a const constructor

**Review by Akanksha-nuclei — CHANGES_REQUESTED** [review]
changes requested

File: Assignment-1/lib/main.dart [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    stdout.write("Enter the item price:\t");
+    try {
+      itemPrice = double.parse(stdin.readLineSync().toString());
+      Util.isGreaterThanZero(itemPrice);
+    } catch (error) {
+      // if we provide other data type or a negative value
+      print("ItemPrice - $error");
+      errorFlag = true;
+    }
+
```
Comment (Akanksha-nuclei) [inline Assignment-1/lib/main.dart:40]: if name is not valid then person should not enter quantity

**Pull request closed without merging**